### PR TITLE
satellite/payments: More Cleanup and Satellite command to ensure we have stripe customers

### PR DESCRIPTION
**Pull request by stefanbenten**
What: 

- Cleans up more stripe code
- Adds a command to the satellite to ensure we have a stripe customer for ever user of the satellite
- Ensures that the Project Limit is approperatily set.
- It refactors our coupon apply function to use config values instead of hardcoded values across the codebase.

Why: We need to ensure those things work.
 
Please describe the performance impact: None

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/payments/paymentsconfig/config.go [hunk 1]
```diff
@@ -4,15 +4,20 @@
...
-	StorageTBPrice     string `help:"price user should pay for storing TB per month" default:"10"`
-	EgressTBPrice      string `help:"price user should pay for each TB of egress" default:"45"`
-	ObjectPrice        string `help:"price user should pay for each object stored in network per month" default:"0.0000022"`
-	BonusRate          int64  `help:"amount of percents that user will earn as bonus credits by depositing in STORJ tokens" default:"10"`
+	StorageTBPrice     string      `help:"price user should pay for storing TB per month" default:"10"`
+	EgressTBPrice      string      `help:"price user should pay for each TB of egress" default:"45"`
+	ObjectPrice        string      `help:"price user should pay for each object stored in network per month" default:"0.0000022"`
+	BonusRate          int64       `help:"amount of percents that user will earn as bonus credits by depositing in STORJ tokens" default:"10"`
+	CouponValue        int64       `help:"coupon value in cents" default:"5500"`
+	CouponDuration     int64       `help:"duration a coupon is valid in months/billing cycles" default:"2"`
```
Comment (littleskunk) [inline satellite/payments/paymentsconfig/config.go]: Please add a hint that his value is only for new coupoons and does not affect any existing coupon.

File: satellite/console/service.go [hunk 2]
```diff
@@ -894,11 +897,19 @@ func (s *Service) CreateProject(ctx context.Context, projectInfo ProjectInfo) (p
 		s.log.Debug(fmt.Sprintf("could not add promotional coupon for user %s", auth.User.ID.String()), zap.Error(Error.Wrap(err)))
 		return p, nil
 	}
-	if len(cards) == 0 {
+
+	balance, err := s.accounts.Balance(ctx, auth.User.ID)
```
Comment (rikysya) [inline satellite/console/service.go]: project seems to be already created at this moment

File: cmd/satellite/main.go [hunk 3]
```diff
@@ -415,6 +422,12 @@ func cmdNodeUsage(cmd *cobra.Command, args []string) (err error) {
 	return generateNodeUsageCSV(ctx, start, end, file)
 }
 
+func cmdStripeCustomer(cmd *cobra.Command, args []string) (err error) {
+	ctx, _ := process.Ctx(cmd)
+
+	return generateStripeCustomers(ctx)
```
Comment (rikysya) [inline cmd/satellite/main.go]: I would make it a grpc call to the satellite

File: satellite/console/service.go [hunk 4]
```diff
@@ -894,11 +897,19 @@ func (s *Service) CreateProject(ctx context.Context, projectInfo ProjectInfo) (p
...
+	if err != nil {
+		s.log.Debug(fmt.Sprintf("could not add promotional coupon for user %s", auth.User.ID.String()), zap.Error(Error.Wrap(err)))
+		return p, nil
+	}
+
+	if len(cards) == 0 && balance < s.minCoinPayment {
```
Comment (rikysya) [inline satellite/console/service.go]: this might work strange when we will enable multiple project per user

Comment (stefanbenten) [reply satellite/console/service.go]: We will rework this entire code after prod launch anyway.

File: cmd/satellite/main.go (hunk 3 above)
Comment (stefanbenten) [reply cmd/satellite/main.go]: This is only a one of command and can likely be removed with the next release.

File: satellite/console/service.go (hunk 2 above)
Comment (stefanbenten) [reply satellite/console/service.go]: Its about applying the proper coupon here, or what are you referring to?

Comment (rikysya) [reply satellite/console/service.go]: If we move fetch of customer balance and credit cards before creating a project, we will prevent project creation when you don't have a card or your balance is too low, or there is an error during retrieving this info so we won't add a coupon. IMHO we want to keep project and coupon in sync as much as we can, right? It's just a suggestion, I am not sure what is the actual requirement

File: cmd/satellite/main.go (hunk 3 above)
Comment (rikysya) [reply cmd/satellite/main.go]: it's applicable to customer payments setup as well, it is more straightforward and easier to manage, as satellite has all appropriate services. Again, that is just a suggestion

File: satellite/console/service.go (hunk 2 above)
Comment (stefanbenten) [reply satellite/console/service.go]: As discussed with @rikysya we are going to address this after this PR in an entire rewrite of it.

File: cmd/satellite/main.go (hunk 3 above)
Comment (stefanbenten) [reply cmd/satellite/main.go]: Going forward each new signup will have a stripe account already. This part only ensures, we have one for users pre billing.

File: cmd/satellite/stripe.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019 Storj Labs, Inc.
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: 2020

File: cmd/satellite/stripe.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// UserData contains the uuid and email of a satellite user
+type UserData struct {
+	ID    uuid.UUID
+	Email string
+}
+
+// generateStripeCustomers creates missing stripe-customers for users in our database
+func generateStripeCustomers(ctx context.Context) error {
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: return type should be named `err error` to use it correctly with defer

File: cmd/satellite/stripe.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	Accounts payments.Accounts
+}
+
+// UserData contains the uuid and email of a satellite user
+type UserData struct {
+	ID    uuid.UUID
+	Email string
+}
+
+// generateStripeCustomers creates missing stripe-customers for users in our database
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: dot at the end of doc message

File: cmd/satellite/stripe.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		return err
+	}
+	if implementation != dbutil.Postgres && implementation != dbutil.Cockroach {
+		return errs.New("unsupported driver %q", driver)
+	}
+
+	dbxDB, err := dbx.Open(driver, source)
+	if err != nil {
+		return errs.New("failed opening database via DBX at %q: %v",
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: can you unwrap this line?

File: cmd/satellite/stripe.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	defer func() {
+		err = errs.Combine(err, dbxDB.Close())
+	}()
+
+	handler, err := setupPayments(zap.L().Named("payments"), db)
+	if err != nil {
+		return err
+	}
+
+	rows, err := dbxDB.Query(ctx, "SELECT id, email FROM users WHERE email NOT IN (SELECT email from stripe_coinpayments")
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: shouldn't this be somewhere in DB code and exposed only by interface method?

File: cmd/satellite/stripe.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	"storj.io/storj/private/dbutil"
+	"storj.io/storj/satellite"
+	"storj.io/storj/satellite/payments"
+	"storj.io/storj/satellite/payments/stripecoinpayments"
+	"storj.io/storj/satellite/satellitedb"
+	"storj.io/storj/satellite/satellitedb/dbx"
+)
+
+// Payments is a wrapper around the Account Handling Service
+type Payments struct {
```
Comment (mniewrzal) [inline cmd/satellite/stripe.go]: Why we need this wrapper?

File: cmd/satellite/stripe.go (hunk 9 above)
Comment (stefanbenten) [reply cmd/satellite/stripe.go]: This is only a short lived tool and i am planning on ripping it out after the next release.

File: satellite/console/service.go [hunk 11]
```diff
@@ -864,6 +867,23 @@ func (s *Service) CreateProject(ctx context.Context, projectInfo ProjectInfo) (p
...
+
+	balance, err := s.accounts.Balance(ctx, auth.User.ID)
+	if err != nil {
+		s.log.Debug(fmt.Sprintf("could not add promotional coupon for user %s", auth.User.ID.String()), zap.Error(Error.Wrap(err)))
+		return nil, Error.Wrap(err)
+	}
+
+	if len(cards) == 0 && balance < s.minCoinPayment {
+		s.log.Debug(fmt.Sprintf("could not add promotional coupon for user %s - no valid payment methods found", auth.User.ID.String()), zap.Error(Error.Wrap(err)))
+		return nil, Error.Wrap(err)
```
Comment (mniewrzal) [inline satellite/console/service.go]: you don't have error here, need to create new one

Comment (stefanbenten) [reply satellite/console/service.go]: Thanks!

Comment (stefanbenten) [reply satellite/console/service.go]: Done

**Pull request merged**